File: algorithms/first_bad_version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):
#     return True if version > 10 else False

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        l, r = 0, n
        while l + 1 < r:
            mid = l + ((r - l) >> 1)
            if isBadVersion(mid):
                r = mid
            else:
                l = mid
        return r

# t = Solution()
# for i in range(11, 100):
#     assert t.firstBadVersion(i) == 11
# print "test passed"

File: algorithms/first_bad_version.rb
# The is_bad_version API is already defined for you.
# @param {Integer} version
# @return {boolean} whether the version is bad
def is_bad_version(version)
  version >= 3
end

# @param {Integer} n
# @return {Integer}
def first_bad_version(n)
  left = 1
  right = n
  while !is_bad_version(left)
    mid = (left + right) / 2
    if is_bad_version(mid)
      right = mid
    else
      left = mid + 1
    end
  end
  left
end

p first_bad_version(10)
